### ISSUE TITLE
Zed refuses to use GPU (Nvidia) WSL 2

### ISSUE THREAD

**Issue by brys0** [zed-industries/zed#15295]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

wsl -v 
```
WSL version: 2.2.4.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.22631.3880
```

Tried the regular ver. of zed as well
zed --version
```
Zed  preview 0.146.0 
```

Attempting to run zed results in it using the llvmpipe software renderer and not my dedicated gpu.

To attempt and resolve this I have tried the following

vkcube
```
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
Selected GPU 0: Microsoft Direct3D12 (NVIDIA GeForce RTX 2080 SUPER), type: 2
``` 

This works, gpu is dedicated correctly and used in the application.

nvidia-smi
```
Fri Jul 26 10:47:57 2024
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 556.12         CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 2080 ...    On  |   00000000:01:00.0  On |                  N/A |
| 22%   55C    P0             61W /  250W |    1796MiB /   8192MiB |      3%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A        35      G   /Xwayland                                   N/A      |
+-----------------------------------------------------------------------------------------+
```

vulkaninfo (truncated)
```
Presentable Surfaces:
=====================
GPU id : 0 (Microsoft Direct3D12 (NVIDIA GeForce RTX 2080 SUPER)):
        Surface types: count = 2
                VK_KHR_xcb_surface
                VK_KHR_xlib_surface
        Formats: count = 2
                SurfaceFormat[0]:
                        format = FORMAT_B8G8R8A8_SRGB
                        colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
                SurfaceFormat[1]:
                        format = FORMAT_B8G8R8A8_UNORM
                        colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
        Present Modes: count = 4
                PRESENT_MODE_IMMEDIATE_KHR
                PRESENT_MODE_MAILBOX_KHR
                PRESENT_MODE_FIFO_KHR
                PRESENT_MODE_FIFO_RELAXED_KHR
 ```

MESA_VK_DEVICE_SELECT=list zed --foreground
```
selectable devices:
  GPU 0: 10de:1e81 "Microsoft Direct3D12 (NVIDIA GeForce RTX 2080 SUPER)" discrete GPU
  GPU 1: 10005:0 "llvmpipe (LLVM 15.0.7, 256 bits)" CPU
```

Things I've tried:

Running zed with `DRI_PRIME=1` 
Running zed with `MESA_VK_DEVICE_SELECT=10de:1e81` (my gpu id)
Running zed with `MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1` (Running zed with this causes the application not to display at all, the logs are similar as well)
Running zed with `VK_ICD_FILENAMES=nvidia_icd.json`
Running zed with `WAYLAND_DISPLAY=''`

Additional steps I've tried:

Restarting computer
Restarting WSL
Reinstalling Ubuntu WSL from scratch
Installing CUDA drivers on Windows and Linux
Reinstalling mesa on Linux

Logs:

### Environment

Processor	Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz   3.60 GHz
Installed RAM	32.0 GB
Device ID	0F91E45C-0632-495F-9E08-DC08052333DE
Product ID	00330-80000-00000-AA933
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>
[Zed_MESA_VK_DEVICE_SELECT&DRI_PRIME.log, Unsupported GPU, but window is displayed](https://github.com/user-attachments/files/16395177/Zed_MESA_VK_DEVICE_SELECT.DRI_PRIME.log)

There are no logs at all when MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1 is set. No logs are created in .local/share/zed/logs/Zed.log

Logs in console are the following:

```log
MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1 MESA_VK_DEVICE_SELECT=10de:1e81 DRI_PRIME=1  WAYLAND_DISPLAY='' zed --foreground

WARNING: dzn is not a conformant Vulkan implementation, testing use only.
```

With `VK_ICD_FILENAMES=nvidia_icd.json` as well logs return an error from `blade_graphics::hal::init`

```
MESA_VK_DEVICE_SELECT_FORCE_DEFAULT_DEVICE=1 MESA_VK_DEVICE_SELECT=10de:1e81 DRI_PRIME=1 VK_ICD_FILENAMES=nvidia_icd.json WAYLAND_DISPLAY='' zed --foreground


[2024-07-26T13:00:06-05:00 ERROR blade_graphics::hal::init] Instance extension "VK_KHR_get_physical_device_properties2" is not supported
```

**Comment by Landon-Kuehner** [comment]
I have also started having this same exact issue on my arch Linux install. Probably NVIDIA causing more issues.

**Comment by king-11** [comment]
does regular zed even work in wsl 2?

**Comment by brys0** [comment]
> does regular zed even work in wsl 2?

No, which is what I said in my issue as well. I tried both regular and preview and neither worked.

**Comment by jonbri** [comment]
In WSL, I was able to get past the error popup with `ZED_ALLOW_EMULATED_GPU=1`

**Comment by zhaoyuheng200** [comment]
> In WSL, I was able to get past the error popup with `ZED_ALLOW_EMULATED_GPU=1`

`ZED_ALLOW_EMULATED_GPU=1` removes the error popup by acknowledging "using a emulated GPU is find by me", but it doesn't solve the issue that zed is using an emulated GPU

**Comment by apestel** [comment]
I think I've been down the rabbit hole...

TLDR; it's not a direct issue with Zed but with Blade the Vulkan library used by Zed which require a Vulkan extension called VK_EXT_inline_uniform_block like OP mentionned in the former post.
Our first hope is a support in Dozen/Mesa for this extension to use our GPU.

Zed is not using your nvidia GPU under WSL2 because the Mesa/Dozen Vulkan implementation which is a passthrough to Direct3D12 lack of VK_EXT_inline_uniform_block extension.

This command show the issue:
VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/dzn_icd.x86_64.json RUST_LOG='blade_graphics=debug' ./zed --foreground .
[2024-09-07T20:26:17+02:00 INFO  blade_graphics::hal::init] Enabling Vulkan Portability
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
[2024-09-07T20:26:18+02:00 INFO  blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (NVIDIA GeForce RTX 4090)"
[2024-09-07T20:26:18+02:00 WARN  blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver!
[2024-09-07T20:26:18+02:00 INFO  blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (AMD Radeon(TM) Graphics)"
[2024-09-07T20:26:18+02:00 WARN  blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver!
error opening [PathWithPosition { path: "/home/adrien/.local/zed.app/bin", row: None, column: None }]: NoSupportedDeviceFound

And now it makes more sense.
If you want to check which driver implement which extension:
$ vulkaninfo

When looking at Blade, the Vulkan library used by Zed, it's written: 
On Vulkan we'll require certain features to make the translation simple:

  - [VK_KHR_push_descriptor](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_push_descriptor.html)
  - [VK_KHR_descriptor_update_template](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_descriptor_update_template.html)
  - [VK_EXT_inline_uniform_block](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_inline_uniform_block.html)
  - [VK_KHR_dynamic_rendering](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_dynamic_rendering.html)

I'm now looking at Dozen which was merged two years ago in Mesa if and when the missing VK_EXT_inline_uniform_block will be implemented. I'll let you know if I find something new. 😅

**Comment by brys0** [comment]
> I think I've been down the rabbit hole...
> 
> TLDR; it's not a direct issue with Zed but with Blade the Vulkan library used by Zed which require a Vulkan extension called VK_EXT_inline_uniform_block like OP mentionned in the former post.
> Our first hope is a support in Dozen/Mesa for this extension to use our GPU.
> 
> Zed is not using your nvidia GPU under WSL2 because the Mesa/Dozen Vulkan implementation which is a passthrough to Direct3D12 lack of VK_EXT_inline_uniform_block extension.
> 
> This command show the issue:
> VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/dzn_icd.x86_64.json RUST_LOG='blade_graphics=debug' ./zed --foreground .
> [2024-09-07T20:26:17+02:00 INFO  blade_graphics::hal::init] Enabling Vulkan Portability
> WARNING: dzn is not a conformant Vulkan implementation, testing use only.
> WARNING: dzn is not a conformant Vulkan implementation, testing use only.
> [2024-09-07T20:26:18+02:00 INFO  blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (NVIDIA GeForce RTX 4090)"
> [2024-09-07T20:26:18+02:00 WARN  blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver!
> [2024-09-07T20:26:18+02:00 INFO  blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (AMD Radeon(TM) Graphics)"
> [2024-09-07T20:26:18+02:00 WARN  blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver!
> error opening [PathWithPosition { path: "/home/adrien/.local/zed.app/bin", row: None, column: None }]: NoSupportedDeviceFound
> 
> And now it makes more sense.
> If you want to check which driver implement which extension:
> $ vulkaninfo
> 
> When looking at Blade, the Vulkan library used by Zed, it's written: 
> On Vulkan we'll require certain features to make the translation simple:
> 
>   - [VK_KHR_push_descriptor](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_push_descriptor.html)
>   - [VK_KHR_descriptor_update_template](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_descriptor_update_template.html)
>   - [VK_EXT_inline_uniform_block](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_inline_uniform_block.html)
>   - [VK_KHR_dynamic_rendering](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_dynamic_rendering.html)
> 
> I'm now looking at Dozen which was merged two years ago in Mesa if and when the missing VK_EXT_inline_uniform_block will be implemented. I'll let you know if I find something new. 😅 

Thank you seriously for all this.

**Comment by chicagobuss** [comment]
If it helps, I'm also getting the emulation in zed on WSL2 with a working Radeon 7900XT.  `vkcube` works fine with all three of my available GPU options, and here is the list output:
```
:~$ MESA_VK_DEVICE_SELECT=list zed --foreground
  GPU 0: 10005:0 "llvmpipe (LLVM 15.0.7, 256 bits)" CPU
  GPU 1: 1002:744c "Microsoft Direct3D12 (AMD Radeon RX 7900 XT)" discrete GPU
  GPU 2: 1002:164e "Microsoft Direct3D12 (AMD Radeon(TM) Graphics)" integrated GPU
```

```
# dedicated GPU
:~$ vkcube --gpu_number 0
Selected GPU 0: Microsoft Direct3D12 (AMD Radeon RX 7900 XT), type: 2

# emulated
:~$ vkcube --gpu_number 1
Selected GPU 1: llvmpipe (LLVM 15.0.7, 256 bits), type: 4

# dedicated integrated GPU
:~$ vkcube --gpu_number 2
Selected GPU 2: Microsoft Direct3D12 (AMD Radeon(TM) Graphics), type: 1
```

zed does work but it doesn't matter which GPU I try to select, it always uses emulation:
## command:
`MESA_VK_DEVICE_SELECT=1002:744c && zed --foreground .`
## result:
![image](https://github.com/user-attachments/assets/32558aef-c975-4d87-bd0c-d02e9ed15386)

**Comment by chiddekel** [comment]
https://github.com/microsoft/wslg/issues/40
Zed developers can contribute to this implementation of MESA:

sudo add-apt-repository ppa:kisak/kisak-mesa
sudo apt update
sudo apt upgrade


That brind d3d deveice to WSL 2 as mesa.
So GPU ZED layer can by develop for WSL2 separately -  for Ubuntu on WSL2.

**Comment by CoderParth** [comment]
@brys0 
Can you please try manually switching to NVIDIA GPU?
`sudo prime-select nvidia`


You seem to have already tried running zed with wayland display. However, could you please try manually switching to nvidia, before running with wayland display?

And then try running zed with `WAYLAND_DISPLAY=''`

**Comment by brys0** [comment]
Hi, unfortunately switching to nvidia with prime-select doesn't work currently on wsl 2 it seems.

```bash
brys0@brys:/mnt/c/Users/bryso$ sudo prime-select nvidia
Error: no integrated GPU detected.
```

**Comment by tkalceca** [comment]
> I think I've been down the rabbit hole...
> 
> TLDR; it's not a direct issue with Zed but with Blade the Vulkan library used by Zed which require a Vulkan extension called VK_EXT_inline_uniform_block like OP mentionned in the former post. Our first hope is a support in Dozen/Mesa for this extension to use our GPU.
> 
> Zed is not using your nvidia GPU under WSL2 because the Mesa/Dozen Vulkan implementation which is a passthrough to Direct3D12 lack of VK_EXT_inline_uniform_block extension.
> 
> This command show the issue: VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/dzn_icd.x86_64.json RUST_LOG='blade_graphics=debug' ./zed --foreground . [2024-09-07T20:26:17+02:00 INFO blade_graphics::hal::init] Enabling Vulkan Portability WARNING: dzn is not a conformant Vulkan implementation, testing use only. WARNING: dzn is not a conformant Vulkan implementation, testing use only. [2024-09-07T20:26:18+02:00 INFO blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (NVIDIA GeForce RTX 4090)" [2024-09-07T20:26:18+02:00 WARN blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver! [2024-09-07T20:26:18+02:00 INFO blade_graphics::hal::init] Adapter: "Microsoft Direct3D12 (AMD Radeon(TM) Graphics)" [2024-09-07T20:26:18+02:00 WARN blade_graphics::hal::init] Rejected for device extension "VK_EXT_inline_uniform_block" not supported. Please update the driver! error opening [PathWithPosition { path: "/home/adrien/.local/zed.app/bin", row: None, column: None }]: NoSupportedDeviceFound
> 
> And now it makes more sense. If you want to check which driver implement which extension: $ vulkaninfo
> 
> When looking at Blade, the Vulkan library used by Zed, it's written: On Vulkan we'll require certain features to make the translation simple:
> 
> * [VK_KHR_push_descriptor](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_push_descriptor.html)
> * [VK_KHR_descriptor_update_template](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_descriptor_update_template.html)
> * [VK_EXT_inline_uniform_block](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_inline_uniform_block.html)
> * [VK_KHR_dynamic_rendering](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_dynamic_rendering.html)
> 
> I'm now looking at Dozen which was merged two years ago in Mesa if and when the missing VK_EXT_inline_uniform_block will be implemented. I'll let you know if I find something new. 😅

Hi @apestel,
Any updates on: " I'm now looking at Dozen which was merged two years ago in Mesa if and when the missing VK_EXT_inline_uniform_block will be implemented. I'll let you know if I find something new."

I see similar issue with my integrated GPU:

```
MESA_VK_DEVICE_SELECT=list zed
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
selectable devices:
  GPU 0: 8086:46a6 "Microsoft Direct3D12 (Intel(R) Iris(R) Xe Graphics)" integrated GPU
  GPU 1: 10005:0 "llvmpipe (LLVM 17.0.6, 256 bits)" CPU
```

**Comment by cpfiffer** [comment]
Same issue here. Zed is only using `llvmpipe`, even though I have an integrated GPU.

```bash
λ ~/ MESA_VK_DEVICE_SELECT=list zed --foreground
MESA-INTEL: warning: cannot initialize blitter engine
selectable devices:
  GPU 0: 8086:7d55 "Intel(R) Arc(tm) Graphics (MTL)" integrated GPU 0000:00:02.0
  GPU 1: 10005:0 "llvmpipe (LLVM 15.0.7, 256 bits)" CPU 0000:00:00.0
```

`.local/share/zed/logs/Zed.log` relevant lines:

```
2025-01-09T11:53:57.999186813-08:00 [INFO] ========== starting zed ==========
2025-01-09T11:53:58.014351383-08:00 [INFO] Compositor detection: _NET_WM_CM_S?=false, _NET_WM_CM_OWNER=false, _NET_SUPPORTING_WM_CHECK=true
2025-01-09T11:53:58.014425109-08:00 [INFO] x11: compositor present: true, gtk_frame_extents_supported: true
2025-01-09T11:53:58.039419662-08:00 [INFO] Presentation support: "VK_KHR_wayland_surface"
2025-01-09T11:53:58.039451572-08:00 [INFO] Presentation support: "VK_KHR_xcb_surface"
2025-01-09T11:53:58.039461185-08:00 [INFO] Presentation support: "VK_KHR_xlib_surface"
2025-01-09T11:53:58.039471629-08:00 [INFO] Enabling Vulkan Portability
2025-01-09T11:53:58.039480149-08:00 [INFO] Enabling color space support
2025-01-09T11:53:58.044162534-08:00 [INFO] Adapter: "Intel(R) Arc(tm) Graphics (MTL)"
2025-01-09T11:53:58.044604203-08:00 [WARN] Rejecting Intel for not presenting when Nvidia is present (on Linux)
2025-01-09T11:53:58.044619754-08:00 [INFO] Adapter: "llvmpipe (LLVM 15.0.7, 256 bits)"
2025-01-09T11:53:58.044665452-08:00 [INFO] No ray tracing extensions are supported

...

2025-01-09T11:53:58.554109933-08:00 [INFO] Using GPU: GpuSpecs { is_software_emulated: true, device_name: "llvmpipe (LLVM 15.0.7, 256 bits)", driver_name: "llvmpipe", driver_info: "Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319 (LLVM 15.0.7)" }
```

Fwiw: the last time my integrated GPU was used is 

```
2025-01-09T09:28:42.936304759-08:00 [INFO] Using GPU: GPUSpecs { is_software_emulated: false, device_name: "Intel(R) Arc(tm) Graphics (MTL)", driver_name: "Intel open-source Mesa driver", driver_info: "Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319" }
```

The only think I can think of changing was my build from Ghostty, which required Zig, gtk4, and a few other things.

**Comment by apestel** [comment]
Just to make things clear: it's not your device, nor your system or your libraries.

Under WSL2, we depend on a Vulkan driver which translates GPU API calls to Direct3D (DirectX 12).
This driver lacks of a feature called [VK_EXT_inline_uniform_block](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_inline_uniform_block.html)
While this feature is missing in this driver, Zed under WSL2 will refuse to launch in a GPU accelerated mode.

You can try to implement this feature in Dozen (Mesa Vulkan to Direct3D Driver). It's out of my knowledge and current capabilities (Person willing to achieve this challenge need to have a good understanding of both Vulkan and Direct3D).

In the meantime, you can build Zed from source. It works well natively under Windows.
There is no "remote" feature currently available in Windows build which is a little bit annoying if you want to work on your WSL2 system like you may do with VSCode for example.

I hope it makes things clearer.

**Comment by tkalceca** [comment]
Ok, thanks. Yeah, I'm using Zed also via Windows (scoop install versions/zed-nightly), but was exactly looking for the "remote" feature and that's why I gave a try on WSL.
Looks like I have to wait for the release on Windows.

**Comment by poe-trade** [comment]
Thanks @apestel - just curious though, is there something Zed could do differently to NOT rely on VK_EXT_inline_uniform_block?

I ask because there are quite a few 3D and accelerated 2D apps that work fine under WSL2 right now.. any ideas?

**Comment by chicagobuss** [comment]
@apestel - thanks for the explanation but can you help us understand _why_ you need that inline_uniform_block call in zed?  Why to apps like google-chrome and vscode run fine in wsl2?  it's not like their gpu performance is bad or anything either.

**Comment by apestel** [comment]
The Zed developers have selected the [Blade](https://github.com/kvark/blade) GPU library to handle the bulk of the rendering workload. This library relies on certain Vulkan features for optimal performance, such as [VK_EXT_inline_uniform_block](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_inline_uniform_block.html).

In contrast, other applications use different GPU rendering implementations and may not require the same Vulkan feature set.

**Comment by chicagobuss** [comment]
Sounds like a... strong choice if this means WSL can't possibly work then, eh?

**Comment by brys0** [comment]
> Sounds like a... strong choice if this means WSL can't possibly work then, eh?
> 
> 

They built blade specifically for rendering in zed. I don't think it'd be a very worthwhile change to create a different renderer

**Comment by isvicy** [comment]
can you have some fallback plan if VK_EXT_inline_uniform_block is only for "optimal performance"? fallback to a normal gpu solution could still be better than fallback to cpu simulation.
